File: src/features/organizations/domain/entities/organization_entity.ts
/**
 * Negocio Flex - Organization & Multi-Tenant Domain Entities (Fase 4)
 * Define las entidades fundamentales de organizaciones, miembros, roles y permisos.
 */

export type BusinessType =
  | 'restaurant'
  | 'salon'
  | 'gym'
  | 'store'
  | 'professional'
  | 'other'
  | 'pasteleria'
  | 'barberia'
  | 'ferreteria'
  | 'veterinaria'
  | 'boutique'
  | 'servicios_generales'
  | 'personalizado';

export type OrganizationStatus = 'active' | 'inactive' | 'suspended';

export type OrganizationRole = 'owner' | 'admin' | 'staff';

export type PlatformRole = 'super_admin' | 'user';

export type MemberStatus = 'active' | 'inactive' | 'invited' | 'suspended';

export type OrganizationAction =
  | 'edit_business_info'
  | 'manage_members'
  | 'change_member_roles'
  | 'remove_members'
  | 'invite_members'
  | 'configure_settings'
  | 'view_metrics'
  | 'view_catalog'
  | 'manage_catalog'
  | 'view_orders'
  | 'manage_orders';

export interface OrgBrandingEntity {
  readonly primaryColor: string;
  readonly secondaryColor: string;
  readonly logoUrl?: string;
  readonly bannerUrl?: string;
  readonly fontName?: string;
  readonly slogan?: string;
}

export interface OrgModulesConfigEntity {
  readonly enableProducts: boolean;
  readonly enableServices: boolean;
  readonly enableAppointments: boolean;
  readonly enableInventory: boolean;
  readonly enableOrders: boolean;
  readonly enableWhatsappCheckout: boolean;
  readonly enableStaffManagement: boolean;
  readonly enableReviews: boolean;
}

export interface OrganizationSettingsEntity {
  readonly id: string;
  readonly organizationId: string;
  readonly language: string;
  readonly timezone: string;
  readonly currency: string;
  readonly dynamicConfig?: Record<string, any>;
  readonly createdAt: string;
  readonly updatedAt: string;
}

export interface OrganizationMemberEntity {
  readonly id: string;
  readonly organizationId: string;
  readonly userId: string;
  readonly role: OrganizationRole;
  readonly status: MemberStatus;
  readonly userFullName?: string;
  readonly userEmail?: string;
  readonly userAvatarUrl?: string;
  readonly createdAt: string;
  readonly updatedAt: string;
}

export interface OrganizationEntity {
  readonly id: string;
  readonly name: string;
  readonly slug: string;
  readonly businessType: BusinessType;
  readonly status: OrganizationStatus;
  readonly createdBy: string;
  readonly createdAt: string;
  readonly updatedAt: string;
  readonly description?: string;
  readonly phone?: string;
  readonly email?: string;
  readonly address?: string;
  readonly currency: string;
  readonly branding: OrgBrandingEntity;
  readonly modules: OrgModulesConfigEntity;
  readonly settings?: OrganizationSettingsEntity;
  readonly currentUserRole?: OrganizationRole;
  readonly memberCount?: number;
}

/**
 * Matriz de Permisos por Rol en la Organización
 */
export const hasPermission = (
  role: OrganizationRole | undefined | null,
  action: OrganizationAction
): boolean => {
  if (!role) return false;

  switch (role) {
    case 'owner':
      // OWNER tiene control total sobre su negocio
      return true;

    case 'admin':
      // ADMIN puede gestionar operaciones y catálogo, pero no eliminar miembros ni cambiar roles de owners
      switch (action) {
        case 'edit_business_info':
        case 'configure_settings':
        case 'view_metrics':
        case 'view_catalog':
        case 'manage_catalog':
        case 'view_orders':
        case 'manage_orders':
        case 'invite_members':
          return true;
        case 'manage_members':
        case 'change_member_roles':
        case 'remove_members':
        default:
          return false;
      }

    case 'staff':
      // STAFF tiene permisos operativos restringidos
      switch (action) {
        case 'view_catalog':
        case 'view_orders':
        case 'manage_orders':
          return true;
        case 'edit_business_info':
        case 'manage_members':
        case 'change_member_roles':
        case 'remove_members':
        case 'invite_members':
        case 'configure_settings':
        case 'view_metrics':
        case 'manage_catalog':
        default:
          return false;
      }

    default:
      return false;
  }
};

File: flutter_project/lib/features/organizations/domain/entities/organization_entity.dart
/// Negocio Flex - Organization Domain Entity (Flutter / Dart)

enum BusinessType {
  restaurant,
  salon,
  gym,
  store,
  professional,
  other,
  pasteleria,
  barberia,
  ferreteria,
  veterinaria,
  boutique,
  serviciosGenerales,
  personalizado,
}

enum OrganizationStatus {
  active,
  inactive,
  suspended,
}

enum OrganizationRole {
  owner,
  admin,
  staff,
}

enum OrganizationAction {
  editBusinessInfo,
  manageMembers,
  changeMemberRoles,
  removeMembers,
  inviteMembers,
  configureSettings,
  viewMetrics,
  viewCatalog,
  manageCatalog,
  viewOrders,
  manageOrders,
}

class OrgBrandingEntity {
  final String primaryColor;
  final String secondaryColor;
  final String? logoUrl;
  final String? bannerUrl;
  final String? fontName;
  final String? slogan;

  const OrgBrandingEntity({
    this.primaryColor = '#4F46E5',
    this.secondaryColor = '#0D9488',
    this.logoUrl,
    this.bannerUrl,
    this.fontName,
    this.slogan,
  });
}

class OrgModulesConfigEntity {
  final bool enableProducts;
  final bool enableServices;
  final bool enableAppointments;
  final bool enableInventory;
  final bool enableOrders;
  final bool enableWhatsappCheckout;
  final bool enableStaffManagement;
  final bool enableReviews;

  const OrgModulesConfigEntity({
    this.enableProducts = true,
    this.enableServices = true,
    this.enableAppointments = false,
    this.enableInventory = true,
    this.enableOrders = true,
    this.enableWhatsappCheckout = true,
    this.enableStaffManagement = true,
    this.enableReviews = true,
  });
}

class OrganizationEntity {
  final String id;
  final String name;
  final String slug;
  final BusinessType businessType;
  final OrganizationStatus status;
  final String createdBy;
  final DateTime createdAt;
  final DateTime updatedAt;
  final String? description;
  final String? phone;
  final String? email;
  final String? address;
  final String currency;
  final OrgBrandingEntity branding;
  final OrgModulesConfigEntity modules;
  final OrganizationRole? currentUserRole;
  final int memberCount;

  const OrganizationEntity({
    required this.id,
    required this.name,
    required this.slug,
    required this.businessType,
    this.status = OrganizationStatus.active,
    required this.createdBy,
    required this.createdAt,
    required this.updatedAt,
    this.description,
    this.phone,
    this.email,
    this.address,
    this.currency = 'S/',
    this.branding = const OrgBrandingEntity(),
    this.modules = const OrgModulesConfigEntity(),
    this.currentUserRole,
    this.memberCount = 1,
  });

  static bool hasPermission(OrganizationRole? role, OrganizationAction action) {
    if (role == null) return false;
    switch (role) {
      case OrganizationRole.owner:
        return true;
      case OrganizationRole.admin:
        switch (action) {
          case OrganizationAction.editBusinessInfo:
          case OrganizationAction.configureSettings:
          case OrganizationAction.viewMetrics:
          case OrganizationAction.viewCatalog:
          case OrganizationAction.manageCatalog:
          case OrganizationAction.viewOrders:
          case OrganizationAction.manageOrders:
          case OrganizationAction.inviteMembers:
            return true;
          case OrganizationAction.manageMembers:
          case OrganizationAction.changeMemberRoles:
          case OrganizationAction.removeMembers:
            return false;
        }
      case OrganizationRole.staff:
        switch (action) {
          case OrganizationAction.viewCatalog:
          case OrganizationAction.viewOrders:
          case OrganizationAction.manageOrders:
            return true;
          default:
            return false;
        }
    }
  }
}
